services/rfq/relayer/service/statushandler.go

```diff
@@ -224,6 +238,103 @@ func (q *QuoteRequestHandler) shouldCheckClaim(request reldb.QuoteRequest) bool
 	return true
 }
 
+// Cache based rate limiter.
+// We don't want to commit a very large balance of the relayer in one block straight away without confirmations.
+// Case 1: If the RFQ demands over $10k in one block, we should not relay the request unless we have a confirmation.
+// Case 2: We should also not relay the request if the cumulative relayed amount over the last blockWindowSize blocks
+// exceeds the volume limit AND we have not confirmed.
+func (q *QuoteRequestHandler) canRelayBasedOnVolumeAndConfirmations(
+	request reldb.QuoteRequest,
+	currentBlockNumber uint64,
+	volumeLimit float64,
+) (bool, error) {
+	// Case 1: Singular RFQ over volumeLimit and inadequate confirmations
+
+	priceOfOriginToken, err := q.getTokenPrice(context.Background(), request)
+	if err != nil {
+		return false, fmt.Errorf("could not get price: %w", err)
+	}
+
+	numOfConfirmations := currentBlockNumber - request.BlockNumber
+
+	if priceOfOriginToken > volumeLimit && numOfConfirmations < 1 {
+		return false, nil
+	}
+
+	if q.rfqCache.Len() > 0 {
+		// Case 2: Cumulative RFQs over volumeLimit and inadequate confirmations
+		blockWindowUSDAmount := q.getBlockWindowRelayedAmount()
+		numOfConfirmations = currentBlockNumber - q.rfqCache.Front().Key
+
+		if blockWindowUSDAmount > volumeLimit && numOfConfirmations < 1 {
+			return false, nil
+		}
+	}
+
+	return true, nil
+}
+
+// addRelayToCache adds the relayed amount to the cache.
+func (q *QuoteRequestHandler) addRelayToCache(ctx context.Context, request reldb.QuoteRequest) error {
+	// If the block number is less than the first block in the window, then we don't need to add it.
+	if q.rfqCache.Front() != nil && request.BlockNumber < q.rfqCache.Front().Key {
+		return nil
+	}
+
+	// Get the token price.
+	priceOfOriginToken, err := q.getTokenPrice(ctx, request)
+	if err != nil {
+		return fmt.Errorf("could not get price: %w", err)
+	}
+
+	// If we have some room, add the RFQ to the cache.
+	cacheLength := q.rfqCache.Len()
+	if cacheLength < q.blockWindowSize {
+		prev := q.rfqCache.GetOrDefault(request.BlockNumber, 0)
+		q.rfqCache.Set(request.BlockNumber, prev+priceOfOriginToken)
```

Comment:
`rfqCache` is meant to store USD volume right? in this case we should be inserting `priceOfOriginToken * request.OriginAmount` (where origin amount is adjusted for decimals)